server/cluster.go

```diff
@@ -0,0 +1,288 @@
+package server
+
+import (
+	"strings"
+	"sync"
+
+	"golang.org/x/net/context"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/pd/protopb"
+)
+
+const (
+	// defaultMaxPeerNumber is the default max peer number for a region.
+	defaultMaxPeerNumber = uint32(3)
+)
+
+var (
+	errClusterNotBootstrapped = errors.New("cluster is not bootstrapped")
+)
+
+// Raft cluster key format:
+// cluster 1 -> /raft/1, value is metapb.Cluster
+// cluster 2 -> /raft/2
+// For cluster 1
+// node 1 -> /raft/1/n/1, value is metapb.Node
+// store 1 -> /raft/1/s/1, value is metapb.Store
+// region 1 -> /raft/1/r/1, value is encoded_region_key
+// region search key map -> /raft/1/rk/encoded_region_key, value is metapb.Region
+//
+// Operation list, pd can only handle operations like auto-balance, split,
+// merge sequentially, and every operation will be assigned a unique incremental ID.
+// pending list -> /raft/1/j/1, /raft/1/j/2, value is operation job.
+//
+// Encode region search key:
+//  1, the maximum end region key is empty, so the encode key is \xFF
+//  2, other region end key is not empty, the encode key is \z end_key
+
+type raftCluster struct {
+	sync.Mutex
+
+	s *Server
+
+	meta        protopb.Cluster
+	clusterRoot string
+
+	// node cache
+	nodes map[uint64]protopb.Node
+	// store cache
+	stores map[uint64]protopb.Store
+}
+
+func (s *Server) newCluster(clusterID uint64, meta protopb.Cluster) *raftCluster {
+	c := &raftCluster{
+		s:           s,
+		meta:        meta,
+		clusterRoot: s.getClusterRootPath(clusterID),
+		nodes:       make(map[uint64]protopb.Node),
+		stores:      make(map[uint64]protopb.Store),
+	}
+
+	return c
+}
+
+func (s *Server) getClusterRootPath(clusterID uint64) string {
+	return strings.Join(
+		[]string{s.cfg.RootPath, "raft", string(uint64ToBytes(clusterID))},
```

Comment:
Maybe you want to use `atoi` instead of `uint64ToBytes`. It may produce invalid chars such as `/`.